File: src/package/vcpkg.rs
use anyhow::{Context, Ok, Result, bail};

use std::{path::PathBuf, process::Command};

use super::PackageInfo;

#[derive(Debug)]
pub struct VcpkgManifest {
    pub exe_root: PathBuf,
}

fn vcpkg_exe() -> Result<PathBuf> {
    let root = crate::core::get_vcpkg_manifest().exe_root.clone();

    let exe = if cfg!(windows) {
        root.join("vcpkg.exe")
    } else {
        root.join("vcpkg")
    };

    if !exe.exists() {
        bail!("vcpkg executable not found at {}", exe.display());
    }

    Ok(exe)
}

pub(super) fn search_package(package: &str) -> anyhow::Result<Vec<super::PackageInfo>> {
    let vcpkg_exe = vcpkg_exe()?;

    let output = Command::new(&vcpkg_exe)
        .args(&["search", package, "--x-full-desc"])
        .output()
        .context("Failed to search packages")?;

    if !output.status.success() {
        bail!("Failed to search for package '{}'", package);
    }

    let output_str = String::from_utf8_lossy(&output.stdout);
    Ok(parse_package_list(&output_str))
}

pub(super) fn install_dependencies(manifest: &crate::core::manifest::Manifest) -> Result<()> {
    let vcpkg_manifest_path = crate::core::get_root().join("vcpkg.json");
    if !vcpkg_manifest_path.exists() {
        let default_vcpkg_json = serde_json::json!({
            "$schema": "https://raw.githubusercontent.com/microsoft/vcpkg-tool/main/docs/vcpkg.schema.json",
            "dependencies": []
        });

        if let Some(parent) = vcpkg_manifest_path.parent() {
            if !parent.exists() {
                std::fs::create_dir_all(parent)?;
            }
        }

        let file_content = serde_json::to_string_pretty(&default_vcpkg_json)?;
        std::fs::write(&vcpkg_manifest_path, file_content)?;
    }

    println!("       Installing dependencies via vcpkg...");

    let vcpkg_exe = vcpkg_exe()?;
    let triplet = Some(crate::util::detect_triplet())
        .ok_or_else(|| anyhow::anyhow!("Triplet not set for vcpkg"))?;

    let status = Command::new(&vcpkg_exe)
        .args(&[
            "install",
            "--x-manifest-root",
            crate::core::get_root()
                .to_str()
                .ok_or_else(|| anyhow::anyhow!("Invalid project root path"))?,
            "--x-install-root",
            &crate::core::get_root()
                .join("packages")
                .to_string_lossy(),
            "--triplet",
            &triplet,
        ])
        .status()
        .context("Failed to run vcpkg install")?;

            if !status.success() {
                bail!("Failed to install dependencies via vcpkg");
            }

    Ok(())
}


fn parse_package_list(output: &str) -> Vec<PackageInfo> {
    let mut packages = Vec::new();

    for line in output.lines() {
        let line = line.trim();
        if line.is_empty() {
            continue;
        }

        let parts: Vec<&str> = line.split_whitespace().collect();

        if parts.len() >= 2 {
            let name = parts[0].to_string();

            if parts[1].chars().next().map_or(false, |c| c.is_digit(10)) {
                let version = parts[1].to_string();
                let description = if parts.len() > 2 {
                    Some(parts[2..].join(" "))
                } else {
                    None
                };

                packages.push(PackageInfo {
                    name,
                    version,
                    description,
                });
            }
        }
    }

    packages
}

pub fn update_vcpkg_manifest(manifest: &mut crate::core::manifest::Manifest) -> Result<()> {
    let vcpkg_manifest_path = crate::core::get_root().join("vcpkg.json");
    let mut vcpkg_json: serde_json::Value = if vcpkg_manifest_path.exists() {
        let content = std::fs::read_to_string(&vcpkg_manifest_path)
            .context("Failed to read vcpkg.json")?;
        serde_json::from_str(&content).context("Failed to parse vcpkg.json")?
    } else {
        serde_json::json!({
            "$schema": "https://raw.githubusercontent.com/microsoft/vcpkg/master/scripts/vcpkg.schema.json",
            "dependencies": []
        })
    };

    let dependencies = vcpkg_json
        .get_mut("dependencies")
        .and_then(|d| d.as_array_mut())
        .ok_or_else(|| anyhow::anyhow!("Invalid vcpkg.json format: 'dependencies' is not an array"))?;

    for (name, version) in &manifest.dependencies {
        if !dependencies.iter().any(|d| d.as_str() == Some(&name)) {
            dependencies.push(serde_json::Value::String(name.clone()));
        }
    }

    let updated_content = serde_json::to_string_pretty(&vcpkg_json)
        .context("Failed to serialize updated vcpkg.json")?;
    std::fs::write(&vcpkg_manifest_path, updated_content)
        .context("Failed to write updated vcpkg.json")?;

    Ok(())
}

File: src/package/vcpkg.cpp
#include "package/vcpkg.hpp"

#include <algorithm>
#include <format>
#include <fstream>
#include <regex>
#include <span>
#include <sstream>
#include <string>
#include <string_view>
#include <unordered_set>
#include <vector>

#include "core/paths.hpp"
#include "util/fs.hpp"
#include "util/process.hpp"
#include "util/result.hpp"
#include "util/text.hpp"

namespace {

auto escape_json_string(std::string_view input) {
    std::string escaped;
    escaped.reserve(input.size() + 2);
    for (char ch : input) {
        if (ch == '\\' || ch == '"') {
            escaped.push_back('\\');
        }
        escaped.push_back(ch);
    }
    return escaped;
}

auto package_install_root(const std::filesystem::path& project_root)
    -> std::filesystem::path {
    return project_root / "packages";
}

auto vcpkg_info_root(const std::filesystem::path& project_root)
    -> std::filesystem::path {
    return package_install_root(project_root) / "vcpkg" / "info";
}

auto normalize_package_name(std::string_view value) -> std::string {
    const auto colon = value.find(':');
    return std::string(value.substr(0, colon));
}

auto path_exists(const std::filesystem::path& path) -> util::Result<bool> {
    std::error_code ec;
    const bool exists = std::filesystem::exists(path, ec);
    if (ec) {
        return std::unexpected(util::make_error(std::format(
            "Failed to access {} ({})", path.string(), ec.message())));
    }
    return exists;
}

auto installed_list_path(const std::filesystem::path& project_root,
                         const package::ResolvedPackage& package,
                         std::string_view triplet) -> std::filesystem::path {
    return vcpkg_info_root(project_root) /
           std::format("{}_{}_{}.list", normalize_package_name(package.name),
                       package.version, triplet);
}

auto package_list_has_installed_entries(
    const std::filesystem::path& install_root,
    const std::filesystem::path& list_file) -> util::Result<bool> {
    std::ifstream input(list_file);
    if (!input.is_open()) {
        return std::unexpected(util::make_error(std::format(
            "Failed to open installed package list: {}", list_file.string())));
    }

    std::string line;
    while (std::getline(input, line)) {
        const auto entry = util::text::trim_ascii_copy(line);
        if (entry.empty()) {
            continue;
        }

        const auto installed_path = install_root / std::filesystem::path(entry);
        const auto exists = GUARD(path_exists(installed_path));
        if (exists) {
            return true;
        }
    }

    return false;
}

auto dependencies_array_bounds(std::string_view json)
    -> util::Result<std::pair<std::size_t, std::size_t>> {
    const auto key = json.find("\"dependencies\"");
    if (key == std::string::npos) {
        return std::unexpected(util::make_error(
            "Invalid vcpkg.json: missing 'dependencies' field."));
    }

    const auto colon = json.find(':', key);
    if (colon == std::string::npos) {
        return std::unexpected(util::make_error(
            "Invalid vcpkg.json: malformed 'dependencies' declaration."));
    }

    const auto open = json.find('[', colon);
    if (open == std::string::npos) {
        return std::unexpected(util::make_error(
            "Invalid vcpkg.json: 'dependencies' is not an array."));
    }

    bool in_string = false;
    bool escaped = false;
    int depth = 0;
    for (std::size_t i = open; i < json.size(); ++i) {
        const char ch = json[i];
        if (escaped) {
            escaped = false;
            continue;
        }
        if (ch == '\\') {
            escaped = true;
            continue;
        }
        if (ch == '"') {
            in_string = !in_string;
            continue;
        }
        if (in_string) {
            continue;
        }
        if (ch == '[') {
            ++depth;
            continue;
        }
        if (ch == ']') {
            --depth;
            if (depth == 0) {
                return std::pair<std::size_t, std::size_t>{open, i};
            }
        }
    }

    return std::unexpected(util::make_error(
        "Invalid vcpkg.json: unterminated dependencies array."));
}

auto collect_dependency_names(const package::ResolvedGraph& graph)
    -> std::vector<std::string> {
    std::vector<std::string> names;
    names.reserve(graph.packages.size());

    std::unordered_set<std::string> seen;
    seen.reserve(graph.packages.size());

    for (const auto& package : graph.packages) {
        if (!seen.insert(package.name).second) {
            continue;
        }
        names.push_back(package.name);
    }

    return names;
}

auto render_dependencies_array(std::span<const std::string> package_names)
    -> std::string {
    if (package_names.empty()) {
        return "[]";
    }

    std::ostringstream out;
    out << "[\n";
    for (std::size_t i = 0; i < package_names.size(); ++i) {
        out << "    \"" << escape_json_string(package_names[i]) << "\"";
        if (i + 1 < package_names.size()) {
            out << ",";
        }
        out << "\n";
    }
    out << "  ]";
    return out.str();
}

}  // namespace

namespace package::vcpkg {

auto vcpkg_executable(const core::Manifest& manifest)
    -> util::Result<std::filesystem::path> {
    if (manifest.features.vcpkg_root.empty()) {
        return std::unexpected(
            util::make_error("vcpkg_root must be set in [features] when "
                             "package_manager is vcpkg."));
    }

#ifdef _WIN32
    const auto exe = manifest.features.vcpkg_root / "vcpkg.exe";
#else
    const auto exe = manifest.features.vcpkg_root / "vcpkg";
#endif

    std::error_code ec;
    if (!std::filesystem::exists(exe, ec) || ec) {
        return std::unexpected(
            util::make_error("vcpkg executable not found: " + exe.string()));
    }
    return exe;
}

auto ensure_vcpkg_manifest(const std::filesystem::path& project_root)
    -> util::Status {
    const auto file = project_root / "vcpkg.json";
    std::error_code ec;
    if (std::filesystem::exists(file, ec) && !ec) {
        return util::Ok;
    }
    if (ec) {
        return std::unexpected(util::make_error(
            "Failed to access vcpkg manifest path: " + file.string()));
    }

    std::ostringstream content;
    content << "{\n";
    content
        << "  \"$schema\": "
           "\"https://raw.githubusercontent.com/microsoft/vcpkg/master/scripts/"
           "vcpkg.schema.json\",\n";
    content << "  \"dependencies\": []\n";
    content << "}\n";

    return util::fs::atomic_write_text_result(file, content.str());
}

auto parse_package_list(std::string_view output)
    -> util::Result<std::vector<PackageInfo>> {
    std::vector<PackageInfo> packages;
    std::istringstream stream{std::string(output)};
    std::string line;

    while (std::getline(stream, line)) {
        line = util::text::trim_ascii_copy(line);
        if (line.empty()) {
            continue;
        }

        std::istringstream row(line);
        std::string name;
        std::string version;
        if (!(row >> name >> version)) {
            continue;
        }
        if (version.empty() ||
            !std::isdigit(static_cast<unsigned char>(version.front()))) {
            continue;
        }

        std::string description;
        std::getline(row, description);
        description = util::text::trim_ascii_copy(description);
        packages.push_back(PackageInfo{name, version, description});
    }

    return packages;
}

auto search_packages(const core::Manifest& manifest, std::string_view package)
    -> util::Result<std::vector<PackageInfo>> {
    auto exe = vcpkg_executable(manifest);
    if (!exe) {
        return std::unexpected(exe.error());
    }

    util::process::RunOptions options;
    options.capture_output = true;
    options.merge_stderr = true;

    const std::vector<std::string> args{"search", std::string(package),
                                        "--x-full-desc"};
    auto result = util::process::run_result(*exe, args, options);
    if (!result) {
        return std::unexpected(result.error());
    }

    if (result->exit_code != 0) {
        const auto details = util::text::trim_ascii_copy(result->output);
        return std::unexpected(util::make_error(
            "Failed to search for package '" + std::string(package) +
            "' via vcpkg.\n" + details));
    }

    return parse_package_list(result->output);
}

auto sync_dependencies(const std::filesystem::path& project_root,
                       const package::ResolvedGraph& graph) -> util::Status {
    const auto file = project_root / "vcpkg.json";
    auto ensure = ensure_vcpkg_manifest(project_root);
    if (!ensure) {
        return std::unexpected(ensure.error());
    }

    std::ifstream input(file);
    if (!input.is_open()) {
        return std::unexpected(util::make_error(
            "Failed to open vcpkg manifest: " + file.string()));
    }

    std::stringstream buffer;
    buffer << input.rdbuf();
    std::string json = buffer.str();
    input.close();

    auto bounds = GUARD(dependencies_array_bounds(json));
    const auto [open, close] = bounds;
    const auto package_names = collect_dependency_names(graph);
    json.replace(open, close - open + 1,
                 render_dependencies_array(package_names));

    return util::fs::atomic_write_text_result(file, json);
}

auto required_packages_installed(const std::filesystem::path& project_root,
                                 const package::ResolvedGraph& graph)
    -> util::Result<bool> {
    if (graph.packages.empty()) {
        return true;
    }

    const auto install_root = package_install_root(project_root);
    const auto triplet = core::detect_triplet();
    const auto triplet_root = install_root / triplet;
    const auto info_root = vcpkg_info_root(project_root);

    const auto triplet_exists = GUARD(path_exists(triplet_root));
    if (!triplet_exists) {
        return false;
    }

    const auto info_exists = GUARD(path_exists(info_root));
    if (!info_exists) {
        return false;
    }

    for (const auto& package : graph.packages) {
        const auto list_file =
            installed_list_path(project_root, package, triplet);
        const auto list_exists = GUARD(path_exists(list_file));
        if (!list_exists) {
            return false;
        }

        const auto has_entries =
            GUARD(package_list_has_installed_entries(install_root, list_file));
        if (!has_entries) {
            return false;
        }
    }

    return true;
}

auto install_dependencies(const std::filesystem::path& project_root,
                          const core::Manifest& manifest,
                          const package::ResolvedGraph& graph) -> util::Status {
    auto ensure = ensure_vcpkg_manifest(project_root);
    if (!ensure) {
        return std::unexpected(ensure.error());
    }

    const auto installed =
        GUARD(required_packages_installed(project_root, graph));
    if (installed) {
        return util::Ok;
    }

    auto exe = vcpkg_executable(manifest);
    if (!exe) {
        return std::unexpected(exe.error());
    }

    const std::vector<std::string> args = {
        std::string{"install"},
        std::string{"--x-manifest-root"},
        project_root.string(),
        std::string{"--x-install-root"},
        package_install_root(project_root).string(),
        std::string{"--triplet"},
        std::string{core::detect_triplet()},
    };

    util::process::RunOptions options{};
    options.capture_output = true;
    options.merge_stderr = true;
    auto result = util::process::run_result(*exe, args, options);
    if (!result) {
        return std::unexpected(result.error());
    }
    if (result->exit_code != 0) {
        const auto details = util::text::trim_ascii_copy(result->output);
        if (details.empty()) {
            return std::unexpected(
                util::make_error("Failed to install dependencies via vcpkg."));
        }
        return std::unexpected(util::make_error(
            "Failed to install dependencies via vcpkg.\n" + details));
    }

    const auto verified =
        GUARD(required_packages_installed(project_root, graph));
    if (!verified) {
        return std::unexpected(
            util::make_error("vcpkg install completed, but required packages "
                             "are still missing from the local install root."));
    }
    return util::Ok;
}

}  // namespace package::vcpkg
